runtime/vm/ContinuationHelpers.cpp

```diff
@@ -90,28 +90,47 @@ createContinuation(J9VMThread *currentThread, j9object_t continuationObject)
 	return result;
 }
 
-void
-synchronizeWithConcurrentGCScan(J9VMThread *currentThread, J9VMContinuation *continuation)
+j9object_t
+synchronizeWithConcurrentGCScan(J9VMThread *currentThread, j9object_t continuationObject, J9VMContinuation *continuation)
 {
+
 	volatile uintptr_t *localAddr = &continuation->state;
-	/* atomically 'or' (not 'set') continuation->state  with currentThread */
-	uintptr_t oldContinuationState = VM_AtomicSupport::bitOr(localAddr, (uintptr_t)currentThread);
+	uintptr_t oldContinuationState = 0;
+	uintptr_t newContinuationState = 0;
+	uintptr_t returnContinuationState = 0;
+	do (
+		oldContinuationState = *localAddr;
+	 	newContinuationState = VM_VMHelpers::setCarrierAndPendingStateToContinuationState(oldContinuationState, currentThread);
+	 	returnContinuationState = VM_AtomicSupport::lockCompareExchange(localAddr, oldContinuationState, newContinuationState);
+	} while (returnContinuationState != oldContinuationState);
+	Assert_VM_false(VM_VMHelpers::isPendingToBeMountedFromContinuationState(returnContinuationState));
+	Assert_VM_Null(VM_VMHelpers::getCarrierThreadFromContinuationState(returnContinuationState));
 
-	Assert_VM_Null(VM_VMHelpers::getCarrierThreadFromContinuationState(oldContinuationState));
+	do {
+		if (VM_VMHelpers::isConcurrentlyScannedFromContinuationState(returnContinuationState)) {
```

Comment:
note that we are in a loop here and we re-read oldValue, so either this should check oldValue or the re-read should store into returnValue